topology.go

```diff
@@ -290,3 +342,69 @@ func GetComponent(ctx context.Context, db *gorm.DB, id string) (*models.Componen
 
 	return &component, nil
 }
+
+type Tag struct {
+	Key string `json:"key"`
+	Val string `json:"val"`
+}
+
+type TopologyResponse struct {
+	componentIDs    []string
+	healthStatusMap map[string]struct{}
+	typeMap         map[string]struct{}
+	tagMap          map[string]struct{}
+
+	Components     models.Components `json:"components,omitempty"`
+	HealthStatuses []string          `json:"healthStatuses,omitempty"`
+	Teams          []string          `json:"teams,omitempty"`
+	Tags           []Tag             `json:"tags,omitempty"`
+	Types          []string          `json:"types,omitempty"`
+}
+
+func (t *TopologyResponse) AddHealthStatuses(s string) {
+	if t.healthStatusMap == nil {
+		t.healthStatusMap = make(map[string]struct{})
+	}
+
+	if _, exists := t.healthStatusMap[s]; !exists {
+		t.HealthStatuses = append(t.HealthStatuses, s)
+		t.healthStatusMap[s] = struct{}{}
+	}
+}
+
+func (t *TopologyResponse) AddType(typ string) {
+	if t.typeMap == nil {
+		t.typeMap = make(map[string]struct{})
+	}
+
+	if _, exists := t.typeMap[typ]; !exists {
+		t.Types = append(t.Types, typ)
+		t.typeMap[typ] = struct{}{}
+	}
+}
+
+func (t *TopologyResponse) AddTag(tags map[string]string) {
+	if t.tagMap == nil {
+		t.tagMap = make(map[string]struct{})
+	}
+
+	for k, v := range tags {
+		tagKey := fmt.Sprintf("%s=%s", k, v)
+		if _, exists := t.tagMap[tagKey]; !exists {
+			t.Tags = append(t.Tags, Tag{Key: k, Val: v})
+			t.tagMap[tagKey] = struct{}{}
+		}
+	}
+}
+
+// populateTopologyResult goes through the components recursively (depth-first)
+// and populates the TopologyResponse struct.
+func populateTopologyResult(components models.Components, res *TopologyResponse) {
+	for _, component := range components {
+		res.componentIDs = append(res.componentIDs, component.ID.String())
+		res.AddTag(component.Labels)
+		res.AddType(component.Type)
+		res.AddHealthStatuses(string(component.Status))
+		populateTopologyResult(component.Components, res)
+	}
+}
```

Comment:
Can we do this at the tree creation level ? We are already iterating through all the components there once, we shouldn't do it twice

Reply:
Tree creation doesn't always happen (_when you need flat structure_) so it's not appropriate to do this at that level. And we're already iterating through components numerous times with all the filters.

Regardless, this might not even be needed once we move to view. So, I'm leaving this as it is.

Reply:
Yeah, doing it in tree creation is not a good idea

I don't think populateTopology function is required anyway, since the query will use `SELECT DISTINCT` for all the unique sets we need